# 4.7 Deriving New Variables

## Importing Libraries

In [36]:
# Import Libraries
import pandas as pd
import numpy as np
import os

## Importing Dataframes

In [37]:
#Pathway to the Dataframes
path = r'C:\Users\Dallas\Desktop\Data Analyst - Projects\Data Immersion\Data Immersion A4\04-2023 Instacart Basket Analysis'

In [38]:
#Importing orders_products_merged.pkl dataframe
df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

## 4.7 Exercise Instructions

In [39]:
#Creating subset of dataframe to avoid processing issues
df = df_ords_prods_merged[:1000000]

In [40]:
df.shape

(1000000, 14)

In [41]:
#If statement to define different price ranges
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [42]:
#Applying the function
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\Dallas\AppData\Local\Temp\ipykernel_20292\3087574758.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [43]:
#Values in the new function
df['price_range'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

In [44]:
#Checking the maxmimum value of the subset dataframe
df['prices'].max()

14.8

In [45]:
#Using the loc() function to refine original funtion; If prices are greater than 15
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\Dallas\AppData\Local\Temp\ipykernel_20292\3658957172.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [46]:
#If prices are greater than 5 and less than or equal to 15
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [47]:
#If prices are less than or equal to 5
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [48]:
#Values in the function
df['price_range_loc'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

In [49]:
#If prices are greater than 15 for df_ords_prods_merged dataframe
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [50]:
#If prices are greater than 5 and less than or equal to 15 for df_ords_prods_merged dataframe
df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [51]:
#If prices are less than or equal to 5 for df_ords_prods_merged dataframe
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [52]:
#Values in the function for df_ords_prods_merged
df_ords_prods_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     21861558
Low-range product     10130750
High-range product      412551
Name: price_range_loc, dtype: int64

In [53]:
df_ords_prods_merged.columns

Index(['order_id', 'user_id', 'order_number', 'order_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', '_merge', 'price_range_loc'],
      dtype='object')

In [54]:
#Find out which day of the week places the most orders
df_ords_prods_merged['order_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: order_day_of_week, dtype: int64

In [55]:
#Create list of which days are the most and least busy using a For Loop
result = []

for value in df_ords_prods_merged["order_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [56]:
#Generated List
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [57]:
#Merge the list with df_ords_prods_merged
df_ords_prods_merged['busiest_day'] = result

In [58]:
df_ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

## 4.7 Exercises

### Question 1. Complete Instructions in the exercise

### Question 2. Amend column to show two busiest days and new column to show two slowest days

In [59]:
#Create new variable to represent the days of the week list
day_count = df_ords_prods_merged['order_day_of_week'].value_counts(dropna = False)

In [60]:
day_count

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: order_day_of_week, dtype: int64

In [61]:
#Shows the index of the day_count df
day_count.index

Int64Index([0, 1, 6, 2, 5, 3, 4], dtype='int64')

In [62]:
#Create a list from the day_count df;  This will allow the amended function to take from the list 
busy_count = list(day_count.index)

In [63]:
busy_count

[0, 1, 6, 2, 5, 3, 4]

In [64]:
result_step2 = []

In [65]:
#Amended function to include the two busiest days and the two slowest days
for value in df_ords_prods_merged["order_day_of_week"]:
  if value == 0 or value == 1:
    result_step2.append("Busiest days")
  elif value == 4 or value == 3:
    result_step2.append("Least busiest days")
  else:
    result_step2.append("Regularly busy")

In [66]:
#Assign the new result to the new column
df_ords_prods_merged['busiest_days'] = result_step2

In [67]:
#Need to determine if the df can merge with the list, checking the size
df_ords_prods_merged["order_day_of_week"].shape

(32404859,)

### Question 3. Check the values of the new column for accuracy

In [68]:
df_ords_prods_merged['busiest_days'].value_counts(dropna = False)

Regularly busy        12916111
Busiest days          11864412
Least busiest days     7624336
Name: busiest_days, dtype: int64

In [69]:
df_ords_prods_merged.shape

(32404859, 17)

### Question 4. Identify the busiest hours of the day.  Labelled by periods of "Most Orders", "Average Orders" ad "Fewest Orders".  Create a new column called "busiest_period_of_day"

In [70]:
#Create a dataframe with the counts for the hours of the day
df_hours = df_ords_prods_merged['order_hour_of_day'].value_counts().to_dict()

In [71]:
df_hours

{10: 2761760,
 11: 2736140,
 14: 2689136,
 15: 2662144,
 13: 2660954,
 12: 2618532,
 16: 2535202,
 9: 2454203,
 17: 2087654,
 8: 1718118,
 18: 1636502,
 19: 1258305,
 20: 976156,
 7: 891054,
 21: 795637,
 22: 634225,
 23: 402316,
 6: 290493,
 0: 218769,
 1: 115700,
 5: 87961,
 2: 69375,
 4: 53242,
 3: 51281}

In [72]:
#determine the quartile bounds of the values in the df_hours df
df_list_percentile = list(df_ords_prods_merged['order_hour_of_day'].value_counts().quantile([0.25, 0.5, 0.75]))

In [73]:
df_list_percentile

[272562.0, 1117230.5, 2556034.5]

### Observations: Anything above 2,556,034.5 is top, which includes hours 10, 11, 14, 15, 13 and 12.  Anything below 272,562 is bottom, whih includes hours 3, 4, 2, 5, 1 and 0.  Anything else is in the middle

In [74]:
#Create a for loop to separate values based on above quartiles
busy_hours = []

In [75]:
for hour in df_ords_prods_merged['order_hour_of_day']:
  if hour in [10, 11, 14, 15, 13, 12]:
    busy_hours.append('Most orders')
  elif hour in [0, 1, 5, 2, 4, 3]:
    busy_hours.append('Fewest orders')
  else:
    busy_hours.append('Average orders')

In [76]:
#Create the busiest_period_of_day column
df_ords_prods_merged['busiest_period_of_day'] = busy_hours

In [77]:
#Checking the output of the two columns
df_ords_prods_merged[['order_hour_of_day', 'busiest_period_of_day']].head()

order_hour_of_day busiest_period_of_day
0                  8        Average orders
1                  7        Average orders
2                 12           Most orders
3                  7        Average orders
4                 15           Most orders

In [78]:
#Checking the size of the column
df_ords_prods_merged['order_hour_of_day'].shape

(32404859,)

In [79]:
df_ords_prods_merged.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day        busiest_days busiest_period_of_day  
0  Regularly busy      Regularly busy        Average orders  
1  Regularly busy  Least busiest days        Average orders  
2  Regularly busy  Least busiest days           Most orders  
3      Least busy  Least busiest days        Average orders  
4      Least busy  Least busiest days           Most orders

### Question 5. Print the frequecy of the new column

In [80]:
#Frequency of the busiest_period_of_day column
df_ords_prods_merged['busiest_period_of_day'].value_counts(dropna = False)

Most orders       16128666
Average orders    15679865
Fewest orders       596328
Name: busiest_period_of_day, dtype: int64

### Question 6. Ensure notebook is clean and structured

### Question 7. Export as a pickle file

In [81]:
#Export as a pickle file
df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merged_update.pkl'))